Update CMake

henricj opened the pull request:
Enhance the [CMake](https://cmake.org/) build to support:

- Building both static and shared libraries (code changes would be needed on Windows to produce a usable DLL).
- [CPack](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Packaging%20With%20CPack.html)
- Building documentation (when possible).
- Configuration checks are done to more closely match the autoconf build.

Two GitHub workflows are supplied to test the CMake-based build.  One does a normal build and the other runs a [CodeQL](https://codeql.github.com/) analysis.  Both workflows run for Windows, Linux, and macOS.

henricj committed:
Add /out/, /.vs/, and CMakeUserPresets.json to .gitignore

henricj committed:
Update the CMake build.
Specify the C++ language version.
Don't let MSVC get misled by user code pages.
Get the CMake build to behave more like the automake build.

henricj committed:
Add CMake to GitHub build.

mywave82 commented:
This also creates version.texi.in :-) Which should be added to configure.ac too later. I can confirm that debian patches Makefile.am to skip doc directory due to it being broken

mywave82 commented:
version.texi.in will with the current autoconf setup generate, nothing is set

```
@at@set UPDATED @LIBBINIO_TEXI_UPDATED@
@at@set UPDATED-MONTH @LIBBINIO_TEXI_MONTH_UPDATED@
@at@set EDITION @libbinio_VERSION@
@at@set VERSION @libbinio_VERSION@
```

mywave82 commented:
Updating configure.ac so that version.text.in works for both, and also suggest that we use ChangeLog instead of doc/version.texi as the source of the date?


``` diff
--- a/configure.ac
+++ b/configure.ac
@@ -1,12 +1,18 @@
 AC_INIT(binio,1.5,[dn.tlp@gmx.net],libbinio)
 AC_CONFIG_SRCDIR(src/binio.cpp)
-AC_CONFIG_FILES([Makefile src/Makefile doc/Makefile src/binio.h libbinio.pc])
+AC_CONFIG_FILES([Makefile src/Makefile doc/Makefile src/binio.h libbinio.pc doc/version.texi])
 AM_INIT_AUTOMAKE
 AM_MAINTAINER_MODE
 #AM_DISABLE_SHARED
 AM_PROG_LIBTOOL
 AC_LANG(C++)
 
+AC_SUBST(at, "@")
+AC_SUBST(libbinio_VERSION, [$VERSION])
+timestamp=`stat --format=%Y ChangeLog`
+AC_SUBST(LIBBINIO_TEXI_UPDATED, [`date -u --date=@$timestamp +"%d %B %Y"`])
+AC_SUBST(LIBBINIO_TEXI_MONTH_UPDATED, [`date -u --date=@$timestamp +"%B %Y"`])
+
 # Check for a sane C/C++ build environment.
 AC_PROG_MAKE_SET
 AC_PROG_INSTALL
```

mywave82 commented:
Second minor issue, following https://alexreinking.com/blog/building-a-dual-shared-and-static-library-with-cmake.html , the dual shared/static install fails, they both overwrite the same file cmake config file during install with oposite static and shared information

```
cmake --install build-shared --prefix _install
...
cmake --install build-static --prefix _install
...
-- Old export file "_install/lib/cmake/libbinio/libbinio-static-targets.cmake" will be replaced.  Removing files [/home/stian/adplug/libbinio-2-build/_install/lib/cmake/libbinio/libbinio-static-targets-release.cmake].
...
```

mywave82 committed:
Without this line, doing a shared build will create

-- Installing: /home/stian/SORT/adplug/libbinio-cmake/_install/lib/cmake/libbinio/libbinio-static-targets.cmake
-- Installing: /home/stian/SORT/adplug/libbinio-cmake/_install/lib/cmake/libbinio/libbinio-static-targets-release.cmake

instead of

-- Installing: /home/stian/SORT/adplug/libbinio-cmake/_install/lib/cmake/libbinio/libbinio-shared-targets.cmake
-- Installing: /home/stian/SORT/adplug/libbinio-cmake/_install/lib/cmake/libbinio/libbinio-shared-targets-release.cmake

if invoked like this:
$ cmake -G Ninja -S . -B build-shared -DBUILD_SHARED_LIBS=YES -DCMAKE_BUILD_TYPE=Release
$ cmake --build build-shared
$ cmake --install build-shared --prefix _install

mywave82 commented:
> Second minor issue, following https://alexreinking.com/blog/building-a-dual-shared-and-static-library-with-cmake.html , the dual shared/static install fails, they both overwrite the same file cmake config file during install with oposite static and shared information
> 
> ```
> cmake --install build-shared --prefix _install
> ...
> cmake --install build-static --prefix _install
> ...
> -- Old export file "_install/lib/cmake/libbinio/libbinio-static-targets.cmake" will be replaced.  Removing files [/home/stian/adplug/libbinio-2-build/_install/lib/cmake/libbinio/libbinio-static-targets-release.cmake].
> ...
> ```

This is fixed with the commit I added on your branch

mywave82 commented:
CI however fails with

```
Error: vcpkg failed with: Could not locate a manifest (vcpkg.json) above the current working directory.
This vcpkg distribution does not have a classic mode instance.



    at Object.<anonymous> (D:\a\_actions\lukka\run-cmake\v10\dist\index.js:6984:23)
    at Generator.next (<anonymous>)
    at fulfilled (D:\a\_actions\lukka\run-cmake\v10\dist\index.js:6942:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Error: run-cmake action execution failed: 'Error: vcpkg failed with: Could not locate a manifest (vcpkg.json) above the current working directory.
This vcpkg distribution does not have a classic mode instance.
```

mywave82 committed:
CMake: workaround for https://github.com/lukka/run-cmake/issues/142
